Documentation/performance/JitOptimizerTodoAssessment.md

```diff
@@ -0,0 +1,132 @@
+difficult to avoid at source level.  The recent work on finally cloning, empty
+finally removal, and empty try removal targeted this.  Writethrough is another
+key optimization here, and we are actively pursuing it.  Other things we've
+discussed include inlining methods with EH and computing funclet callee-save
+register usage independently of main function callee-save register usage, but
+I don't think we have any particular data pointing to either as a high priority.
+
+
+Loop Optimizations
+------------------
+
+We haven't been targeting benchmarks that spend a lot of time doing compuations
+in an inner loop.  Pursuing loop optimizations for the peanut butter effect
+would seem odd.  So this simply hasn't bubbled up in priority yet, though it's
+bound to eventually.
+
+
+More Expression Optimizations
+-----------------------------
+
+We again don't have particular benchmarks pointing to key missing cases, and
+balancing the CQ vs TP will be delicate here, so it would really help to have
+an appropriate benchmark suite to evaluate this work against.
+
+
+Forward Substitution
+--------------------
+
+This too needs an appropriate benchmark suite that I don't think we have at
+this time.  The tradeoffs against register pressure increase and throughput
+need to be evaluated.  This also might make more sense to do if/when we can
+handle SSA renames.
+
+
+Value Number Conservativism
+---------------------------
+
+We have some frustrating phase-ordering issues resulting from this, but the
+opt-repeat experiment indicated that they're not prevalent enough to merit
+pursuing changing this right now.  Also, using SSA def as the proxy for value
+number would require handling SSA renaming, so there's a big dependency chained
+to this.
+Maybe it's worth reconsidering the priority based on throughput?
+
+
+High Tier Optimizations
+-----------------------
+
+We don't have that many knobs we can "crank up" (though we do have the tracked
+assertion count and could switch inliner policies), nor do we have any sort of
+benchmarking story set up to validate whether tiering changes are helping or
+hurting.  We should get that benchmarking story sorted out and at least hook
+up those two knobs.
+
+
+Low Tier Back-Off
+-----------------
+
+We have some changes we know we want to make here: morph does more than it needs
+to in minopts, and tier 0 should be doing throughput-improving inlines, as
+opposed to minopts which does no inlining.  It would be nice to have the
+benchmarking story set up to measure the effect of such changes when they go in,
+we should do that.
+
+
+Async
+-----
+
+We've made note of the prevalence of async/await in modern code (and particularly
+in web server code such as TechEmpower), but haven't really identified concrete
+opportunities presented here, to my knowledge.  Some sort of study of async
+peanut butter is probably in order, but what would that look like?
```

Comment:
> ECMA IL spec only requires Release semantic for stores with IL volatile prefix

CLR has traditionally implemented a stronger memory model than that ECMA requires. Where "traditionally" means x86/x64 and "implemented" means "the JIT has not implemented any kind of store optimizations". Those 2 coupled give you a MM where all stores have release semantics (modulo holes in x86/x64's MM). If ARM64 can get away without this then that's great (though I'm sure some code will be broken by this 😄).

That said, the problem with `await` is quite a bit more complicated. The `struct` starts its life on the stack so up to `await` the JIT could do whatever it wants with field loads and stores. Stores to fields must be issues if `IsCompleted` is `false`, otherwise they're no required provided that the JIT can reload the fields on the resume path. But how is the JIT supposed to know all that? In particular, understanding that the resume path is taken when `m_state != 0` seems rather impossible unless some hints are provided to the JIT.